src/core/observer/watcher.js

```diff
@@ -230,30 +230,29 @@ export default class Watcher {
  * is collected as a "deep" dependency.
  */
 const seenObjects = new Set()
-function traverse (val: any, seen?: Set) {
+function traverse (val: any) {
+  seenObjects.clear()
+  _traverse(val, seenObjects)
+}
+
+function _traverse (val: any, seen: Set) {
   let i, keys
-  if (!seen) {
-    seen = seenObjects
-    seen.clear()
-  }
   const isA = Array.isArray(val)
-  const isO = isObject(val)
-  if ((isA || isO) && Object.isExtensible(val)) {
+  if ((isA || isObject(val)) && Object.isExtensible(val)) {
```

Comment:
Thanks @HerringtonDarkholme !
Looks Good! 👍 

Do you mean that `Object.isExtensible(val)` is not necessary?
(Sorry, I am not good at English :) )